src/binding/c/attr_api.txt

```diff
@@ -0,0 +1,229 @@
+MPI_Keyval_free:
+    .desc: Frees an attribute key for communicators
+    .seealso: MPI_Keyval_create, MPI_Comm_free_keyval
+    .replace: deprecated with MPI_Comm_free_keyval
+
+MPI_Comm_create_keyval:
+    .desc: Create a new attribute key
+    .extra: AttrErrReturn
+    .seealso: MPI_Comm_free_keyval
+/*
+    Notes:
+    Key values are global (available for any and all communicators).
+
+    Default copy and delete functions are available.  These are
+    + MPI_COMM_NULL_COPY_FN   - empty copy function
+    . MPI_COMM_NULL_DELETE_FN - empty delete function
+    - MPI_COMM_DUP_FN         - simple dup function
+
+    There are subtle differences between C and Fortran that require that the
+    copy_fn be written in the same language from which 'MPI_Comm_create_keyval'
+    is called.
+    This should not be a problem for most users; only programmers using both
+    Fortran and C in the same program need to be sure that they follow this rule.
+*/
+
+MPI_Comm_delete_attr:
+    .desc: Deletes an attribute value associated with a key on a communicator
+    .seealso: MPI_Comm_set_attr, MPI_Comm_create_keyval
+    .extra: ignore_revoke
+
+MPI_Comm_free_keyval:
+    .desc: Frees an attribute key for communicators
+{
+    MPIR_free_keyval(comm_keyval_ptr);
+    *comm_keyval = MPI_KEYVAL_INVALID;
+}
+
+MPI_Comm_get_attr:
+    .desc: Retrieves attribute value by key
+    .polymorph: MPIR_Attr_type attr_type=MPIR_ATTR_PTR
+    .skip: Fortran
+    .extra: ignore_revoke
+/*
+    Notes:
+        Attributes must be extracted from the same language as they were inserted
+        in with 'MPI_ATTR_PUT'.  The notes for C and Fortran below explain why.
+*/
+/*
+    Notes for C:
+        Even though the 'attribute_val' argument is declared as 'void *', it is
+        really the address of a void pointer (i.e., a 'void **').  Using
+        a 'void *', however, is more in keeping with C idiom and allows the
+        pointer to be passed without additional casts.
+
+    .N Fortran
+        The 'attribute_val' in Fortran is a pointer to a Fortran integer, not
+        a pointer to a 'void *'.
+*/
+
+MPI_Comm_set_attr:
+    .desc: Stores attribute value associated with a key
+    .seealso: MPI_Comm_create_keyval, MPI_Comm_delete_attr
+    .polymorph: MPIR_Attr_type attr_type=MPIR_ATTR_PTR
+    .extra: ignore_revoke
+/*
+    Notes:
+    Values of the permanent attributes 'MPI_TAG_UB', 'MPI_HOST', 'MPI_IO',
+    'MPI_WTIME_IS_GLOBAL', 'MPI_UNIVERSE_SIZE', 'MPI_LASTUSEDCODE', and
+    'MPI_APPNUM' may not be changed.
+
+    The type of the attribute value depends on whether C, C++, or Fortran
+    is being used.
+    In C and C++, an attribute value is a pointer ('void *'); in Fortran, it is an
+    address-sized integer.
+
+    If an attribute is already present, the delete function (specified when the
+    corresponding keyval was created) will be called.
+*/
+
+MPI_Type_create_keyval:
+    .desc: Create an attribute keyval for MPI datatypes
+    .extra: AttrErrReturn
```

Comment:
Minor comment: can we make these flags a bit more readable? This makes reviewers easy to review the PR (especially to check which keyword affects which part). Regarding this PR:
1.  `AttrErrReturn`
   If it only affects the comment, how about `DocAttrErrReturn` or using a different `.xxx` (say `.doc: AttrErrReturn `)?
2. `ignore_revoke`
   Is there a rule for capitalization? For example, should it be  `IgnoreRevoke`? I appreciate if the keyword contains more information (e.g., `IgnoreRevokedComm`).

Reply:
`AttrErrReturn` are the exact phrase defined in `maint/docnotes` and in the output as `.N AttrErrReturn`.

`.doc: ` is a good idea!

I don't really like camel case. How about `ignore_revoked_comm`? 

But it is still cryptic as you may still don't understand what does ignore_revoked_comm means. In the end, all these directives are cryptic -- either you know what it means and what you want or don't. The alternative is to directly write custom "error_check" code block. The reason we opt for directives is for brevity and easier maintenance in the long run (when maintainer understands what these directives means or knows how to look them up). So I am hesitate for the attempt to make the word more descriptive.

Reply:
> `.doc`

That's great!

> But it is still cryptic as you may still don't understand what does ignore_revoked_comm means.

I understand, but that does not mean we can keep `ignore_revoke`. I believe at least we should include domains (in this case, maybe "error" and/or "comm") in the keyword. I don't think making such a keyword short makes maintenance easier. A non-maintainer (like reviewers of the potential future PRs) cannot understand it.

I think `ignore_revoked_comm` is much better.  Perhaps putting `error` would make it more readable.

Reply:
I'll work on it (add `.doc` and `ignore_revoked_comm`)

Reply:
Added `.docnotes:` and renamed to `ignore_revoked_comm`